### PR TITLE
add holder for mutex

### PR DESCRIPTION
**Pull request by flyingfish89**
## Summary

add holder for mutex, continue the work https://github.com/apache/nuttx/pull/7426

## Impact

remove holder from rmutex and add holder for mutex

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by masayuki2009** [comment]
@flyingfish89 

I noticed that sim:usrsocktest causes assert with this PR?

```
nsh> uname -a
NuttX  12.0.0 a4c9da9280 Jan 18 2023 20:32:01 sim sim
nsh> usrsocktest
Starting unit-tests...
Testing group "char_dev" =>
	Group "char_dev": [OK]
Testing group "no_daemon" =>
	Group "no_daemon": [OK]
Testing group "basic_daemon" =>
	Group "basic_daemon": [OK]
Testing group "basic_connect" =>
	Group "basic_connect": [OK]
Testing group "basic_connect_delay" =>
	Group "basic_connect_delay": [OK]
Testing group "no_block_connect" =>
	Group "no_block_connect": [OK]
Testing group "basic_send" =>
	Group "basic_send": [OK]
Testing group "no_block_send" =>
	Group "no_block_send": [OK]
Testing group "block_send" =>
	Group "block_send": [OK]
Testing group "no_block_recv" =>
	Group "no_block_recv": [OK]
Testing group "block_recv" =>
	Group "block_recv": [OK]
Testing group "remote_disconnect" =>
	Group "remote_disconnect": [OK]
Testing group "basic_setsockopt" =>
	Group "basic_setsockopt": [OK]
Testing group "basic_getsockopt" =>
	Group "basic_getsockopt": [OK]
Testing group "basic_getsockname" =>
	Group "basic_getsockname": [OK]
Testing group "wake_with_signal" =>
_assert: Current Version: NuttX  12.0.0 a4c9da9280 Jan 18 2023 20:32:01 sim
_assert: Assertion failed at file: usrsock/usrsock_devif.c:709 task: <pthread> 0x5619e6fec5ad
```

File: include/semaphore.h [hunk 1]
```diff
@@ -42,6 +42,7 @@
 #define SEM_PRIO_INHERIT          1
 #define SEM_PRIO_PROTECT          2
 #define SEM_PRIO_MASK             3
+#define SEM_PRIO_MUTEX            4
```
Comment (pkarashchenko) [inline include/semaphore.h:45]:
```suggestion
#define SEM_TYPE_MUTEX            4
```

**Comment by masayuki2009** [comment]
@flyingfish89 

>I noticed that sim:usrsocktest causes assert with this PR.

It seems that this PR causes deadlock with ostest (e.g. rp2040:smp with CONFIG_DEBUG_ASSERTIONS=y).

**Comment by xiaoxiang781216** [comment]
> @flyingfish89
> 
> I noticed that sim:usrsocktest causes assert with this PR.
> 
> ```
> nsh> uname -a
> NuttX  12.0.0 a4c9da9280 Jan 18 2023 20:32:01 sim sim
> nsh> usrsocktest
> Starting unit-tests...
> Testing group "char_dev" =>
> 	Group "char_dev": [OK]
> Testing group "no_daemon" =>
> 	Group "no_daemon": [OK]
> Testing group "basic_daemon" =>
> 	Group "basic_daemon": [OK]
> Testing group "basic_connect" =>
> 	Group "basic_connect": [OK]
> Testing group "basic_connect_delay" =>
> 	Group "basic_connect_delay": [OK]
> Testing group "no_block_connect" =>
> 	Group "no_block_connect": [OK]
> Testing group "basic_send" =>
> 	Group "basic_send": [OK]
> Testing group "no_block_send" =>
> 	Group "no_block_send": [OK]
> Testing group "block_send" =>
> 	Group "block_send": [OK]
> Testing group "no_block_recv" =>
> 	Group "no_block_recv": [OK]
> Testing group "block_recv" =>
> 	Group "block_recv": [OK]
> Testing group "remote_disconnect" =>
> 	Group "remote_disconnect": [OK]
> Testing group "basic_setsockopt" =>
> 	Group "basic_setsockopt": [OK]
> Testing group "basic_getsockopt" =>
> 	Group "basic_getsockopt": [OK]
> Testing group "basic_getsockname" =>
> 	Group "basic_getsockname": [OK]
> Testing group "wake_with_signal" =>
> _assert: Current Version: NuttX  12.0.0 a4c9da9280 Jan 18 2023 20:32:01 sim
> _assert: Assertion failed at file: usrsock/usrsock_devif.c:709 task: <pthread> 0x5619e6fec5ad
> ```

I just tried sim:usrsocktest with the last master(c19b5eeb8d8f37394c79ff85876c7ba6d846fc12), all test can pass.

**Comment by masayuki2009** [comment]
@xiaoxiang781216 

>I just tried sim:usrsocktest with the last master(https://github.com/apache/nuttx/commit/c19b5eeb8d8f37394c79ff85876c7ba6d846fc12), all test can pass.

Please try the test with `CONFIG_DEBUG_ASSERTIONS=y`

**Comment by masayuki2009** [comment]
@flyingfish89 
spresense:rndis_smp is unstable now (i.e. ping does not work)

**Comment by masayuki2009** [comment]
Shall we revert this PR because there are some issues?

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216
> 
> > I just tried sim:usrsocktest with the last master([c19b5ee](https://github.com/apache/nuttx/commit/c19b5eeb8d8f37394c79ff85876c7ba6d846fc12)), all test can pass.
> 
> Please try the test with `CONFIG_DEBUG_ASSERTIONS=y`

Ok, I can repro the problem now. @masayuki2009 could you try this patch: https://github.com/apache/nuttx/pull/8164

File: include/semaphore.h (hunk 1 above)
Comment (xiaoxiang781216) [reply include/semaphore.h:45]: @pkarashchenko here  is the patch: https://github.com/apache/nuttx/pull/8182/commits/d45ac2b157e7b9e486708ac375a44b21c4d90303

**Pull request merged**